src/Network/Request.php

```diff
@@ -1942,13 +1943,27 @@ public function getUri()
      * and `url` attributes.
      *
      * @param \Psr\Http\Message\UriInterface $uri The new request uri
+     * @param bool $preserveHost Whether or not the host should be retained.
      * @return static
      */
-    public function withUri(UriInterface $uri)
+    public function withUri(UriInterface $uri, $preserveHost = false)
     {
         $new = clone $this;
         $new->uri = $uri;
 
+        if ($preserveHost && $this->hasHeader('Host')) {
+            return $new;
+        }
+
+        $host = $uri->getHost();
+        if (!$host) {
+            return $new;
+        }
+        if ($uri->getPort()) {
```

Comment:
The `Uri` object generally ignores the port. We'd also have to ignore 443 when the scheme is https.